WireAPI/Sources/WireAPI/Models/Backend/ServerTrustValidator.swift

```diff
@@ -0,0 +1,112 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+@preconcurrency import Security
+
+struct ServerTrustValidator: Sendable {
+
+    enum Failure: Error {
+        case cannotLoadCertificate
+        case cannotRetrievePublicKey
+        case evaluatingServerTrustFailed((any Swift.Error)?)
+        case noPublicKeyOnServerTrust
+        case noMatchingPublicKey
+    }
+
+    private let pinnedKeys: [PinnedKey]
+
+    init(pinnedKeys: [PinnedKey]) {
+        self.pinnedKeys = pinnedKeys
+    }
+
+    /// Verifies the server `trust` for the given `host`.
+    ///
+    /// - Parameter trust: The `SecTrust` of the server.
+    /// - Parameter host: The host of the server.
+    /// - Throws: An error if server certificate should not be trusted.
+    /// - Note: If no pinned keys are found for the `host`, the server certificate is trusted.
+
+    func validate(trust: SecTrust, host: String) async throws {
+        let matchingKeys = pinnedKeys.filter { $0.matches(host: host) }
+
+        // If no keys are pinned for `host`, we trust the server certificate
+        guard !matchingKeys.isEmpty else { return }
+
+        try await Self.verifyServerCertificateTrusted(trust)
+
+        let publicKey = try Self.publicKeyAssociatedWithServerTrust(trust)
+
+        let publicKeys = try matchingKeys.map { try Self.certificateKey(for: $0) }
+        guard publicKeys.contains(publicKey) else {
+            throw Failure.noMatchingPublicKey
+        }
+    }
+
+    // MARK: - Private
+
+    private static func verifyServerCertificateTrusted(_ serverTrust: SecTrust) async throws {
```

Comment:
I see some methods are static inside the `ServerTrustValidator` do they need to be static ?

Reply:
I personally prefer using `static` where possible because if I do that I know it is likely fairly independent and not relying on instance variables. I almost like to default to `static`. Do you know any cons to this?

Reply:
I was afraid this could lead to some confusion since this method can be called on the type itself and it seems here we only call it from an instance method, not a big fan of mixing up instance and type methods when we don't need to tbh however I see your point and it makes sense too.

Reply:
I see. I won't change now but lets discuss in one of the Architecture forums.